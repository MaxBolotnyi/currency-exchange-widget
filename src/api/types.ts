export type TRateRequestParams = {
    amount: string | number;
    from: string;
    to: string;
};
