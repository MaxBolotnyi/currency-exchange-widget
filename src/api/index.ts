import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import type { TRateRequestParams } from './types';


// Define a service using a base URL and expected endpoints
export const conversionApi = createApi({
    reducerPath: 'conversion',
    baseQuery: fetchBaseQuery({ baseUrl: 'https://api.exchangerate.host/' }),
    endpoints: (builder) => ({
        getConversionRates: builder.query<any, TRateRequestParams>({
            query: ({ amount, from, to }: TRateRequestParams) => `convert?from=${from}&to=${to}&amount=${amount}`,
            transformResponse: (response: any) => ({
                rate: response?.info?.rate,
                amount: response?.result
            })
        })
    }),
});

// auto-generated based on the defined endpoints
export const { useGetConversionRatesQuery } = conversionApi;
