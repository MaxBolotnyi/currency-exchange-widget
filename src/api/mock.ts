import React from 'react';

const getUseMockApi = (success = true) => {
    return ({
        amount,
        from,
        to
    }: any, {
        skip = false,
        polling
    }: any) => {
        const [data, setData] = React.useState({ rate: 1.23, amount: 234.34 });

        const refetch = () => {
            setData({
                rate: Math.random() * 2,
                amount: 100 * Math.random() * 20
            });
        }
        return {
            isError: !success,
            isSuccess: success,
            data: success ? data : undefined,
            refetch
        };
    };
};

export default getUseMockApi;
