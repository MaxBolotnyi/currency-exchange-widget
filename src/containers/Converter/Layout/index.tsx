import React from 'react';

import FormHelperText from '@material-ui/core/FormHelperText';

import AccountSelect from '../../../components/AccountSelect';
import CurrencyInput from '../../../components/CurrencyInput';
import TopToast from '../../../components/TopToast';
import Card from '../../../components/Card';
import Button from '../../../components/Button';

import { useStyles } from './styles';
import { getCurrencySymbol } from '../../../utils/currencies';

import type { TConverterLayoutProps } from './types';


const ConverterLayout = React.memo(({
    showToast = false,
    onToastClose,
    accounts = [],
    src,
    dest,
    isValid = false,
    exchangeRate,
    onSumbit
}: TConverterLayoutProps) => {
    const classes = useStyles();

    const outOfBound = +src.balance < +src.amount;
    return (
        <Card>
            <TopToast
                severity="success"
                open={showToast}
                onDelayedClose={onToastClose}
            >
                Done!
        </TopToast>
            <div className={classes.container}>
                <AccountSelect
                    id="sourceAcc"
                    name="sourceAcc"
                    accounts={accounts}
                    value={src.accountId}
                    className={classes.select}
                    label="Source Account"
                    onChange={src.onAccountChange}
                />
                <AccountSelect
                    id="destAcc"
                    name="destAcc"
                    value={dest.accountId}
                    accounts={accounts}
                    className={classes.select}
                    label="Destination Account"
                    onChange={dest.onAccountChange}
                />
                <div className={classes.row}>
                    <CurrencyInput
                        id="srcAmount"
                        name="srcAmount"
                        error={outOfBound}
                        helperText={outOfBound && 'Not enough money'}
                        className={classes.marginRight}
                        currency={src.currency}
                        value={src.amount}
                        onChange={src.onValChange}
                    />
                    <CurrencyInput
                        id="destAmount"
                        name="destAmount"
                        currency={dest.currency}
                        value={dest.amount}
                        onChange={dest.onValChange}
                    />
                </div>
            </div>
            {
                !outOfBound
                && !!src.accountId
                && !!dest.accountId
                && !!src.currency
                && !!dest.currency
                && !!src.amount
                && !!dest.amount
                && src.currency !== dest.currency
                && (
                    <FormHelperText className={classes.rateInfo}>
                        {!exchangeRate ? `Loading...` : `Rate: ${getCurrencySymbol(src.currency)}1 = ${getCurrencySymbol(dest.currency)}${exchangeRate}`}
                    </FormHelperText>
                )}
            <Button
                disabled={!isValid}
                onClick={onSumbit}>
                Exchange
        </Button>
        </Card >
    )
});

export default ConverterLayout;
