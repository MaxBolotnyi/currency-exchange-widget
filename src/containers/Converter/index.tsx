import React from 'react';

import ConverterLayout from './Layout';
import useConversionForm from '../../hooks/useConversionForm';
import useAccounts from '../../hooks/useAccounts';

import { useGetConversionRatesQuery } from '../../api';

import type { TAccountSelectOnChangeProps } from '../../components/AccountSelect/types';
import type { TCurrencyInputOnChangeProp } from '../../components/CurrencyInput/types';


const Converter = React.memo(() => {
    const [showSuccessToast, setShowSuccessToast] = React.useState(false);
    const {
        state,
        list,
        getById
    } = useAccounts();

    const {
        sourceAcc,
        sourceAmount,
        destAcc,
        destAmount,
        setSourceAcc,
        setDestAcc,
        setSrcAmount,
        setDestAmount,
        isValid,
        exchangeRate,
        onSubmit
    } = useConversionForm({ accounts: state, useApiHook: useGetConversionRatesQuery });

    const onSourceAccChange = (event: TAccountSelectOnChangeProps) => {
        setSourceAcc(event.target.value);
    };

    const onDestAccChange = (event: TAccountSelectOnChangeProps) => {
        setDestAcc(event.target.value);
    };

    const onSourceAmountChange = (event: TCurrencyInputOnChangeProp) => {
        setSrcAmount(event.value ? String(event.value) : '');
    };

    const onDestAmountChange = (event: TCurrencyInputOnChangeProp) => {
        setDestAmount(event.value ? String(event.value) : '');
    };

    const onExchange = () => {
        onSubmit();
        setShowSuccessToast(true);
    };

    const hideToast = React.useCallback(() => {
        setShowSuccessToast(false)
    }, []);

    return (
        <ConverterLayout
            accounts={list}
            isValid={isValid}
            showToast={showSuccessToast}
            onToastClose={hideToast}
            onSumbit={onExchange}
            exchangeRate={exchangeRate}
            src={{
                accountId: sourceAcc,
                currency: getById(sourceAcc)?.currency,
                amount: sourceAmount,
                balance: getById(sourceAcc)?.balance,
                onAccountChange: onSourceAccChange,
                onValChange: onSourceAmountChange
            }}
            dest={{
                accountId: destAcc,
                currency: getById(destAcc)?.currency,
                amount: destAmount,
                onAccountChange: onDestAccChange,
                onValChange: onDestAmountChange
            }}
        />
    );
});

export default Converter;
