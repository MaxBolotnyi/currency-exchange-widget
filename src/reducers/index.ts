import { combineReducers } from '@reduxjs/toolkit';

import { conversionApi } from '../api';
import { accountsSlice } from '../reducers/accounts';

export const rootReducer = combineReducers({
    accounts: accountsSlice.reducer,
    [conversionApi.reducerPath]: conversionApi.reducer,
});
