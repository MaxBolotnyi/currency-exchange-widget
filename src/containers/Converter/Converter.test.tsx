import React from 'react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { rootReducer } from '../../reducers';
import { conversionApi } from '../../api';

import { render } from '@testing-library/react';
import Converter from './index';


const store = configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware().concat(conversionApi.middleware),
});

const { container } = render(<Provider store={store}>
    <Converter />
</Provider>)

test('matches the snapshot', () => {
    expect(container).toMatchSnapshot();
});
