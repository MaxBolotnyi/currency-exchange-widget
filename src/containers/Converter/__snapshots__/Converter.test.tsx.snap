// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<div>
  <div
    class="makeStyles-container-6"
  >
    <div
      class="MuiPaper-root MuiAlert-root MuiAlert-filledSuccess makeStyles-root-7  MuiPaper-elevation0"
      role="alert"
    >
      <div
        class="MuiAlert-icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4C12.76,4 13.5,4.11 14.2, 4.31L15.77,2.74C14.61,2.26 13.34,2 12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0, 0 22,12M7.91,10.08L6.5,11.5L11,16L21,6L19.59,4.58L11,13.17L7.91,10.08Z"
          />
        </svg>
      </div>
      <div
        class="MuiAlert-message"
      >
        Done!
      </div>
    </div>
    <div
      class="makeStyles-container-4"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-select-2"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="sourceAcc"
          id="sourceAcc-label"
        >
          Source Account
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="sourceAcc-label sourceAcc"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
            id="sourceAcc"
            role="button"
            tabindex="0"
          >
            <span>
              ​
            </span>
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            name="sourceAcc"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-11"
            >
              <span>
                Source Account
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-select-2"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="destAcc"
          id="destAcc-label"
        >
          Destination Account
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="destAcc-label destAcc"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
            id="destAcc"
            role="button"
            tabindex="0"
          >
            <span>
              ​
            </span>
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            name="destAcc"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-11"
            >
              <span>
                Destination Account
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="makeStyles-row-3"
      >
        <div
          class="MuiFormControl-root makeStyles-marginRight-1"
        >
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart"
            >
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionStart"
                data-testid="currency-sign"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                >
                  $
                </p>
              </div>
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart"
                disabled=""
                id="srcAmount"
                name="srcAmount"
                placeholder="0.00"
                style="text-align: right;"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-10"
                  style="width: 0.01px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiFormControl-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart"
            >
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionStart"
                data-testid="currency-sign"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                >
                  $
                </p>
              </div>
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart"
                disabled=""
                id="destAmount"
                name="destAmount"
                placeholder="0.00"
                style="text-align: right;"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-10"
                  style="width: 0.01px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root makeStyles-root-13 MuiButton-text Mui-disabled makeStyles-disabled-14 Mui-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label makeStyles-label-15"
      >
        Exchange
      </span>
    </button>
  </div>
</div>
`;
